app/decorators/service_provider_session_decorator.rb

```diff
@@ -39,10 +40,14 @@ def sp_name
   end
 
   def sp_return_url
-    sp.return_to_sp_url
+    if (request_url = session[:request_url])
```

Comment:
tiny style thing: I don't usually see parens around `if` statements in ruby. it is fine to include them, just not something I usually see style-wise